zebra-state/src/in_memory.rs

```diff
@@ -0,0 +1,119 @@
+use super::{Request, Response};
+use futures::prelude::*;
+use std::{
+    collections::{BTreeMap, HashMap},
+    error::Error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+#[derive(Default)]
+struct ZebraState {
+    by_hash: HashMap<BlockHeaderHash, Arc<Block>>,
+    by_height: BTreeMap<BlockHeight, Arc<Block>>,
+}
+
+impl ZebraState {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) {
```

Comment:
This works well for now, but eventually we will need to insert and get by `Transaction` and/or `TransactionHash`s too. We can decide if we want to explicitly namespace them or just use typing to distinguish them, but maybe `tip()` should be explicitly `get_block_tip()`? I'm undecided. @hdevalence wdyt?

Reply:
I think if this were split out into a `BlockIndex` substructure this problem ould go away right?

Reply:
Ah yes, this is solved by the different `Request` enum variants. 👍